### ISSUE TITLE
#272 broke new TS 4.1+ typing features

### ISSUE THREAD

**Issue by tec27** [molefrog/wouter#291]
The changes in #272 seem to break the new typing features (e.g. automatically extracting types from route strings). According to the TS docs on `typeVersions`:

https://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html#multiple-fields

> Since ranges have the potential to overlap, determining which redirect applies is order-specific.

By moving `>=4.1` below a `*` match string, the 4.1+ case will never be hit (even if no files inside the `*` block are being looked for). Testing locally, reversing the order of these blocks makes things work on my machine again.

This seems like a possibly simple fix (just reverse the order of these blocks, duplicate the content of the `*` block in the `>=4.1` block as well), but as I'm not sure what issue this change was solving in the first place/how to test it, I don't really feel comfortable making that change myself.

**Comment by molefrog** [comment]
I did a simple experiment and sadly reversing the order didn't help. It worked for the main entry, but it broke types of other files such as `wouter/use-location`. I don't have the expertise to say for sure why it didn't work, but perhaps we could use a more bulletproof approach here:

```json
  "typesVersions": {
    ">=4.1": {
      "types/ts3.9.4/index.d.ts": [
        "types/ts4.1/index.d.ts"
      ],
      "use-location": [
        "types/use-location.d.ts"
      ],
      "matcher": [
        "types/matcher.d.ts"
      ],
      "static-location": [
        "types/static-location.d.ts"
      ]
    },
    "<4.1": {
      "use-location": [
        "types/use-location.d.ts"
      ],
      "matcher": [
        "types/matcher.d.ts"
      ],
      "static-location": [
        "types/static-location.d.ts"
      ]
    }
  },
```

**Comment by vicero** [comment]
I just ran into this issue as well! :heavy_plus_sign: :one:

**Comment by molefrog** [comment]
Hey @vicero, I've published a test version with the fix applied `2.10.2-dev.1`
Could you please check if it solves your problem?
